src/spatial/flatten.jl

```diff
@@ -0,0 +1,96 @@
+end
+
+"""
+convert to Float64 if type Any
+"""
+function flatten(netstoch::AbstractArray{Any}, reactstoch::AbstractArray{Any}, rx_rates::Vector{Any}, spatial_system, u0::Matrix{Int}, tspan, hopping_constants::Vector{Matrix{F}}; kwargs...) where F <: Number
+    net_stoch = convert(Vector{Vector{Pair{Int64, Int64}}}, netstoch)
+    react_stoch = convert(Vector{Vector{Pair{Int64, Int64}}}, reactstoch)
+    rates = convert(Vector{Float64}, rx_rates)
+    flatten(net_stoch, react_stoch, rates, spatial_system, u0, tspan, hopping_constants; kwargs...)
+end
+
+"""
+"flatten" the spatial jump problem. Return flattened DiscreteProblem and MassActionJump.
+"""
+function flatten(netstoch::Vector{R}, reactstoch::Vector{R}, rx_rates::Matrix{F}, spatial_system, u0::Matrix{Int}, tspan, hopping_constants::Vector{Matrix{F}}; scale_rates = true, kwargs...) where {R, F <: Number}
+    num_species = size(u0, 1)
+    num_nodes = num_sites(spatial_system)
+    num_rxs = length(reactstoch)
+    @assert length(hopping_constants) == size(u0, 2) == size(rx_rates, 2) == num_nodes
+    @assert length(netstoch) == length(reactstoch) == size(rx_rates, 1) == num_rxs
+    spec_CI = CartesianIndices((num_species, num_nodes))
+    spec_LI = LinearIndices((num_species, num_nodes))
+
+    sum_outdegrees = sum(num_neighbors(spatial_system, site) for site in 1:num_nodes)
+    num_jumps = num_species*(sum_outdegrees + num_rxs)
+    total_netstoch = R[]; sizehint!(total_netstoch, num_jumps)
+    total_reactstoch = R[]; sizehint!(total_reactstoch, num_jumps)
+    total_rates = F[]; sizehint!(total_rates, num_jumps)
+
+    #hops
+    # assuming hopping_constants isa Vector{Matrix} where hopping_constants[src][species, index] is the hop const of species from src to neighbor at index in neighbors(spatial_system, src)
+    for src in 1:num_nodes
+        hopping_constants_at_src = hopping_constants[src]
+        for (i,dst) in enumerate(neighbors(spatial_system, src))
+            for species in 1:num_species
+                dst_idx = spec_LI[spec_CI[species,dst]]
+                src_idx = spec_LI[spec_CI[species,src]]
+                push!(total_netstoch, [dst_idx => 1, src_idx => -1])
+                push!(total_reactstoch, [src_idx => 1])
+                push!(total_rates, hopping_constants_at_src[species, i])
+            end
+        end
+    end
+
+    #reactions
+    for site in 1:num_nodes
+        append!(total_netstoch, map(rx_stoch -> map(p -> Pair(spec_LI[spec_CI[p[1],site]], p[2]), rx_stoch), netstoch))
+        append!(total_reactstoch, map(rx_stoch -> map(p -> Pair(spec_LI[spec_CI[p[1],site]], p[2]), rx_stoch), reactstoch))
+    end
+    append!(total_rates, vec(rx_rates)) # assuming rx_rates isa Matrix where rx_rates[rx, site] is the rate of rx at site
+
+    # put everything together
+    ma_jump = MassActionJump(total_rates, total_reactstoch, total_netstoch; nocopy = true, scale_rates = scale_rates)
```

Comment:
If `total_rates` are coming from a user-supplied `MassActionJump`, I think we can assume that they were already scaled in creating it. So here `scale_rates=false` is probably what we want. Sorry for any confusion about this!

Reply:
The method on line 7 (https://github.com/SciML/DiffEqJump.jl/pull/192/files/a330c42c4a8909b909753c8bc32005369201de8f#diff-208a047f849b0efd9cf21d343ed068871f368976846397bc992f78ce89512f9bR7) is the one that takes in a `MassActionJump`. It then calls the main `flatten` method (https://github.com/SciML/DiffEqJump.jl/pull/192/files/a330c42c4a8909b909753c8bc32005369201de8f#diff-208a047f849b0efd9cf21d343ed068871f368976846397bc992f78ce89512f9bR54) with `scale_rates = false`.